cmd/juju/action/run.go

```diff
@@ -141,6 +143,19 @@ func (c *runCommand) Init(args []string) (err error) {
 		return errors.New("no action specified")
 	}
 
+	applicationName, _, foundUnitNumber := strings.Cut(c.unitReceivers[0], "/")
```

Comment:
Just use `names.UnitApplication()` to get the application name from the unit name
You can use this in the loop above - add the app names to a `collections.Set` and then above where the other checks are done, just check that the set size is 1. No need for a separate loop below.

Reply:
`names.UnitApplication()` supports only the real unit name syntax (`name/1`), not the extra `name/leader` shorthand, so I can't use it directly.  afaict `names.UnitApplication()` should support only the pure definition of a unit name and not know the `/leader` 

Instead, I've stolen the method used by `names.UnitApplication` but applied it on a regex that'll match both regular and the special leader unit.  wdyt?